components/causal_ts/src/hlc.rs

```diff
@@ -0,0 +1,159 @@
+        let before = self.0.fetch_max(to.into_inner(), Ordering::Acquire).into();
+        debug!("Hlc::advance"; "before" => ?before, "after" => ?std::cmp::max(before, to));
+    }
+}
+
+const TSO_REFRESH_INTERVAL: Duration = Duration::from_millis(500);
+
+/// A causal timestamp provider by HLC
+pub struct HlcProvider {
+    hlc: Arc<Hlc>,
+    pd_client: Arc<dyn PdClient>,
+    tso_worker: Worker,
+
+    /// Interval of Physical Clock refresh from TSO.
+    /// Smaller interval will get more accurate metrics, but more pressure on PD.
+    /// Can be Duration::ZERO, which means never refresh from TSO. For test purpose ONLY.
+    tso_refresh_interval: Duration,
+
+    initialized: AtomicBool,
+}
+
+impl HlcProvider {
+    pub fn new(pd_client: Arc<dyn PdClient>) -> Self {
+        Self::new_opt(pd_client, TSO_REFRESH_INTERVAL)
+    }
+
+    pub fn new_opt(pd_client: Arc<dyn PdClient>, tso_refresh_interval: Duration) -> Self {
+        Self {
+            hlc: Arc::new(Hlc::default()),
+            pd_client,
+            tso_worker: WorkerBuilder::new("hlc_tso_worker").create(),
+            tso_refresh_interval,
+            initialized: AtomicBool::new(false),
+        }
+    }
+
+    pub async fn init(&self) -> Result<()> {
+        let tso = self.pd_client.get_tso().await?;
+        self.hlc.advance(tso);
+        self.initialized.store(true, Ordering::Release);
+        debug!("HlcProvider::init"; "tso" => ?tso);
+
+        let hlc = self.hlc.clone();
+        let pd_client = self.pd_client.clone();
+
+        let task = move || {
+            let hlc = hlc.clone();
+            let pd_client = pd_client.clone();
+
+            async move {
+                match pd_client.get_tso().await {
+                    Ok(tso) => {
+                        debug!("HlcProvider::pd_client::get_tso"; "tso" => ?tso);
+                        hlc.advance(tso);
+                    }
+                    Err(err) => {
+                        warn!("HlcProvider::pd_client::get_tso error"; "error" => ?err);
+                        // Fail to acquire TSO do not violate correctness. Only affect some duration metrics of CDC.
+                    }
+                };
+            }
+        };
+
+        // Duration::ZERO means never refresh from TSO. For test purpose ONLY.
+        if self.tso_refresh_interval > Duration::ZERO {
+            self.tso_worker
+                .spawn_interval_async_task(self.tso_refresh_interval, task);
+        }
+        Ok(())
+    }
+}
+
+impl CausalTsProvider for HlcProvider {
+    fn get_ts(&self) -> Result<TimeStamp> {
+        if !self.initialized.load(Ordering::Acquire) {
+            Err(Error::Hlc("Uninitialized".to_string()))
+        } else {
+            Ok(self.hlc.next())
+        }
+    }
+
+    fn advance(&self, to: TimeStamp) -> Result<()> {
+        self.hlc.advance(to);
+        Ok(())
+    }
+}
+
+/// A causal timestamp provider by HLC for TEST.
+#[derive(Default)]
+pub struct TestHlcProvider {
```

Comment:
Why not use `Hlc` directly?

Reply:
En... then `Hlc` should implement `CausalTsProvider` trait, which seems to be a little strange as it is also a member of `HlcProvider`.
Becides, `TestHlcProvider` is for test purpose, and I think there will be more methods in the future.